src/Agent.Worker/Container/ContainerInfo.cs

```diff
@@ -144,11 +207,124 @@ public MountVolume(string sourceVolumePath, string targetVolumePath, bool readOn
             this.SourceVolumePath = sourceVolumePath;
             this.TargetVolumePath = targetVolumePath;
             this.ReadOnly = readOnly;
+            this.IsExpanded = true;
+        }
+
+        public MountVolume(string raw, bool isExpanded)
+        {
+            this.IsExpanded = isExpanded;
+            if (this.IsExpanded)
+            {
+                ParseVolumeString(raw);
+            }
+            else
+            {
+                this.Raw = raw;
+            }
+        }
+
+        private void ParseVolumeString(string volume)
+        {
+            var volumeSplit = volume.Split(":");
+            if (volumeSplit.Length == 3)
+            {
+                // source:target:ro
+                SourceVolumePath = volumeSplit[0];
+                TargetVolumePath = volumeSplit[1];
+                ReadOnly = String.Equals(volumeSplit[2], "ro", StringComparison.OrdinalIgnoreCase);
+            }
+            else if (volumeSplit.Length == 2)
+            {
+                if (String.Equals(volumeSplit[1], "ro", StringComparison.OrdinalIgnoreCase))
+                {
+                    // target:ro
+                    TargetVolumePath = volumeSplit[0];
+                    ReadOnly = true;
+                }
+                else
+                {
+                    // source:target
+                    SourceVolumePath = volumeSplit[0];
+                    TargetVolumePath = volumeSplit[1];
+                    ReadOnly = false;
+                }
+            }
+            else
+            {
+                // target - or, default to passing straight through
+                TargetVolumePath = volume;
+                ReadOnly = false;
+            }
         }
 
         public string SourceVolumePath { get; set; }
         public string TargetVolumePath { get; set; }
         public bool ReadOnly { get; set; }
+        public bool IsExpanded { get; set; }
+        public string Raw { get; set; }
```

Comment:
is `IsExpanded` and `Raw` used publicly?

Reply:
Used in JobRunner for delayed variable expansion

Reply:
Removed this, instead using UserPortMappings and UserMountVolumes in ContainerInfo to represent values given in yaml. These are Dictionary<string, string> that can be expanded directly